gem/optimise.py

```diff
@@ -260,6 +287,78 @@ def count(pair):
     return product, flops
 
 
+def associate_sum(summands):
+    """Apply associativity rules to construct an operation-minimal summation tree.
+
+    For best performance give factors that have different set of free indices.
+    """
+    # group summands by their free indices
+    groups = OrderedDict()
+    for summand in summands:
+        key = frozenset(summand.free_indices)
+        groups.setdefault(key, []).append(summand)
+    summands = [reduce(Sum, _value, Zero()) for _value in itervalues(groups)]
+
+    if len(summands) > 32:
+        # O(N^3) algorithm
+        raise NotImplementedError("Not expected such a complicated expression!")
+
+    def count(pair):
+        """Operation count to multiply a pair of GEM expressions"""
+        a, b = pair
+        extents = [i.extent for i in set().union(a.free_indices, b.free_indices)]
+        return numpy.prod(extents, dtype=int)
```

Comment:
Copied code, in fact, this function is basically identical to `associate_product` except that you glue the factors together with `Sum`, rather than `Product`.  But then, thinking about it, is this flop count even correct for sum trees?  Given that the docstring talks about multiplication, I would say no.